### PR TITLE
doc: error and progress logging for information.hpp

### PR DESCRIPTION
**Pull request by nathanvaughan-NOAA**
# What is the feature?

Add FIMS_LOG outputs to monitor progress and errors in information.hpp
* Added progress and error logging to information.hpp

# How have you implemented the solution?

Added lots of FIMS_LOG calls throughout CreateModel() function

# Does the PR impact any other area of the project?

No

## Input

None

## Output

Reports more details to the existing fims.log file.

# How to test this change

No new tests

# Developer pre-PR checklist

- [x] I relied on GitHub actions to :test_tube: things for me while I sat on the :couch_and_lamp:.
- [x] Ran [cmake build and ctest locally](https://noaa-fims.github.io/collaborativ/e_workflow/testing.html#c-unit-testing-and-benchmarking) and the C++ tests passed
- [x] Ran `devtools::check()` locally and the package compiled and all R tests passed. If there are failing tests, run `devtools::test(filter = "file_name")` (where "test-file_name.R" is the file containing the failing tests) to troubleshoot tests.

### CODE REVIEW

**Review by Cole-Monnahan-NOAA — CHANGES_REQUESTED** [review]
I was able to successfully build and test this PR locally. It outputs to fims.log in the root directory, but because output is appended it is difficult to parse and check functionality b/c its unclear if errors are expected or not. The only mandatory changes I suggest are to fix address location printing. A few suggestions for discussion/consideration:

1. File should not append from previous runs. How will this work when we run multiple FIMS models in parallel?
2. The file is very hard to read. I suggest more white space to indent like regions (e.g., all likelihood checks for a given fleet), and breaks between major sections (e.g., selex and growth). 
3. I mentioned it below but I don't think printing the whole data set to the log is necessarily a good idea. I suggest we consider just printing the first and last 3 observations. It's possible something can go wrong in other years, but the user won't detect that from this file. Instead, I think that FIMS should check for invalid individual likelihood components as it calculates them and be able to error and with informative messages like "nll for age comps for fleet 1 year 35 is invalid" then print the observed and expected before exiting.

File: inst/include/common/information.hpp [hunk 1]
```diff
@@ -156,141 +156,168 @@ class Information {
...
       if (f->observed_index_data_id != -999) {
         uint32_t index_id = static_cast<uint32_t>(f->observed_index_data_id);
         data_iterator it = this->data_objects.find(index_id);
-        FIMS_LOG << index_id << std::endl;
+        FIMS_LOG << "Input fleet index id = " << index_id << "." << std::endl;
 
         if (it != this->data_objects.end()) {
           f->observed_index_data = (*it).second;
+          FIMS_LOG << "Index data successfully set." << std::endl;
+          FIMS_LOG << "Observed input index: \n " << f->observed_index_data
```
[the remark this answers is not in the document]
Comment (nathanvaughan-NOAA) [reply inst/include/common/information.hpp]: Fixed this it now just prints the first value. We would need extra code in the future to loop and print more/all of the data.

File: inst/include/common/information.hpp [hunk 2]
```diff
@@ -156,141 +156,168 @@ class Information {
...
+          << std::endl;
         } else {
           valid_model = false;
-          // log error
-          FIMS_LOG << "Error: observed index data not defined for fleet"
-                   << f->id << std::endl;
+          FIMS_LOG << "Error: Expected data observations not defined for fleet"
```
[the remark this answers is not in the document]
Comment (nathanvaughan-NOAA) [reply inst/include/common/information.hpp]: Good point I agree. I didn't catch that. I have added exit calls in all failure cases.

File: inst/include/common/information.hpp [hunk 3]
```diff
@@ -156,141 +156,168 @@ class Information {
...
-        FIMS_LOG << " number of data objects" << this->data_objects.size()
-                 << std::endl;
-
+        FIMS_LOG << "Input fleet age comp id = " << agecomp_id
+               << "." << std::endl;
+        
         if (it != this->data_objects.end()) {
           f->observed_agecomp_data = (*it).second;
-          FIMS_LOG << "expected input agecomp " << f->observed_agecomp_data;
+          FIMS_LOG << "Age comp data successfully set." << std::endl;
```
[the remark this answers is not in the document]
Comment (nathanvaughan-NOAA) [reply inst/include/common/information.hpp]: At the moment I've fallen back to just printing the first value. We will need looping code to print out more and I agree it's questionable what we even want to print. Personally, I only use the SS data logs to see when the model has read something different than what I expected to identify a missing conditional input or similar issue. Given the way FIMS is setup I don't think this should ever be a problem so it's unlikely this is needed. We can discuss options at the meeting next week though.

[the remark this answers is not in the document]
Comment (nathanvaughan-NOAA) [reply inst/include/common/information.hpp]: I have also set up the logging to send logs to their own folder/subfolders. An interesting benefit I found was that if the folder doesn't exist then the log files are not created. We need to do some timed tests to see the impact on runtime but hopefully this is a simple way for the user to determine which if any log files to create we could also add flags to the input code but that will take more code/effort.

**Review by ChristineStawitz-NOAA — APPROVED** [review]

**Comment by kellijohnson-NOAA** [comment]
Given that this is just a branch and not the main trunk, I would suggest that the commit history be fixed to just wipe out the last four commits and roll it back to [34617f5](https://github.com/NOAA-FIMS/FIMS/pull/424/commits/34617f5d3a53f4219079d4ddb334420ef00e6c2e). This can easily be done with 
```
git reset --h 34617f5
git push -f origin 407-information-logs
```
or through an interactive rebase which is easier with git lens where you can just choose to drop the last four commits. I am happy to do either option or walk someone through the steps so that others can learn. But in my opinion there is no reason to push four commits that we essentially do not want into main.

**Comment by nathanvaughan-NOAA** [comment]
Thanks Cole, in response to comments:
1) File should not append from previous runs. How will this work when we run multiple FIMS models in parallel?
a) The file is no longer appending it is just running 4 iterations of create_model() one for each AD level.

2) The file is very hard to read. I suggest more white space to indent like regions (e.g., all likelihood checks for a given fleet), and breaks between major sections (e.g., selex and growth).
a) Agree with this, I have pulled most of the other things out of the file but still need to figure out how to do useful indenting when each line of the file is written independently?

3) I mentioned it below but I don't think printing the whole data set to the log is necessarily a good idea. I suggest we consider just printing the first and last 3 observations. It's possible something can go wrong in other years, but the user won't detect that from this file. Instead, I think that FIMS should check for invalid individual likelihood components as it calculates them and be able to error and with informative messages like "nll for age comps for fleet 1 year 35 is invalid" then print the observed and expected before exiting.
a) For now I have modified this to just print the first value of the data file. We will need to create loops to print multiple values or nans but I agree that that is probably unneeded overhead at the moment.

**Comment by ChristineStawitz-NOAA** [comment]
@Cole-Monnahan-NOAA see Nathan's response - I think you need to approve that he has addressed your comments before we can merge this in. 

We are planning to fix the issue of four runs being printed after we merge in pull request #453 since the Rcpp refactor will change where it needs to be written to the log.

and thanks @kellijohnson-NOAA for giving me the emotional support and git syntax I needed to do a hard reset.

**Review by Cole-Monnahan-NOAA — APPROVED** [review]

**Pull request merged**